packages/pluggableWidgets/accordion-web/src/components/AccordionGroup.tsx

```diff
@@ -0,0 +1,50 @@
+import { createElement, Dispatch, ReactElement, ReactNode, useCallback, useEffect, useRef, useState } from "react";
+
+import { AccordionGroupsReducerAction } from "../utils/AccordionGroupStateReducer";
+
+export interface AccGroup {
+    header: ReactNode;
+    content: ReactNode;
+    collapsed?: boolean;
+    visible: boolean;
+}
+
+export interface AccordionGroupProps {
+    group: AccGroup;
+    accordionGroupsDispatch?: Dispatch<AccordionGroupsReducerAction>;
+}
+
+export default function AccordionGroup(props: AccordionGroupProps): ReactElement | null {
+    const { group, accordionGroupsDispatch } = props;
+
+    const previousVisiblePropValue = useRef(group.visible);
+    const [divContentMounted, setDivContentMounted] = useState(group.visible && !group.collapsed);
+
+    useEffect(() => {
+        if (group.visible !== previousVisiblePropValue.current || !group.collapsed) {
+            previousVisiblePropValue.current = group.visible;
+            setDivContentMounted(group.visible && !group.collapsed);
+        }
+    }, [group, setDivContentMounted]);
+
+    const toggleContentVisibility = useCallback(() => {
+        if (group.collapsed) {
+            accordionGroupsDispatch!({ type: "expand", group });
+        } else {
+            accordionGroupsDispatch!({ type: "collapse", group });
+        }
+    }, [group, accordionGroupsDispatch]);
```

Comment:
Regarding the `onToggle`: I would just remove this callback alltogether and pull it up to `Accordion.tsx`. Right now this component is not optimized for re-render anyways because `group` is an object and not memoized (the array is, but not this object). I would suggest just breaking up all the properties from the `AccGroup` object and make them individual APIs, so:
```
export interface AccordionGroupProps {
    header: ReactNode;
    content: ReactNode;
    collapsed?: boolean;
    visible: boolean;
    onToggle: (isCollapsed: boolean) => void;
}
```

I feel like from a React design perspective this also makes more sense for its composability. Then in the parent you can just do `onToggle={(isCollapsed) => accordionGroupsDispatch({ type: isCollapsed ? "expand" : "collapse", group })}`. I don't think performance or memoization should really be a concern here since the tree doesn't go too deep after this. If you really wanted you could memoize per group, but that might be quite overkill.

Reply:
I agree with all of the above. I will make sure to include this in a story on our backlog to improve.

Reply:
[Jira issue](https://mendix.atlassian.net/browse/WC-506)